recipes/mpfr/all/conanfile.py

```diff
@@ -29,148 +34,153 @@ class MpfrConan(ConanFile):
 
     def build_requirements(self):
-        if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
-            self.build_requires("msys2/cci.latest")
+        if self._settings_build.os == "Windows":
+            self.win_bash = True
+            if not self.conf.get("tools.microsoft.bash:path", check_type=str):
+                self.tool_requires("msys2/cci.latest")
+
+    def layout(self):
+        if self.settings.os == "Windows":
+            cmake_layout(self, src_folder="src")
+        else:
+            basic_layout(self, src_folder="src")
 
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version],
-                  destination=self._source_subfolder, strip_root=True)
-
-    def _configure_autotools(self):
-        if self._autotools:
-            return self._autotools
-        self._autotools = AutoToolsBuildEnvironment(self, win_bash=tools.os_info.is_windows)
-        yes_no = lambda v: "yes" if v else "no"
-        args = [
-            "--enable-thread-safe",
-            "--with-gmp-include={}".format(tools.unix_path(os.path.join(self.deps_cpp_info[str(self.options.exact_int)].rootpath, "include"))),
-            "--with-gmp-lib={}".format(tools.unix_path(os.path.join(self.deps_cpp_info[str(self.options.exact_int)].rootpath, "lib"))),
-            "--enable-shared={}".format(yes_no(self.options.shared)),
-            "--enable-static={}".format(yes_no(not self.options.shared)),
-        ]
+        get(self, **self.conan_data["sources"][self.version],
+            destination=self.source_folder, strip_root=True)
+
+    def generate(self):
+        if self.settings.os == "Windows":
+            if is_msvc(self):
+                # Use NMake to workaround bug in MSBuild versions prior to 2022 that shows up as:
+                #    error MSB6001: Invalid command line switch for "cmd.exe". System.ArgumentException: Item
+                #                   has already been added. Key in dictionary: 'tmp'  Key being added: 'TMP'
+                self.conf.define("tools.cmake.cmaketoolchain:generator", "NMake Makefiles")
+            tc = CMakeToolchain(self)
+            tc.generate()
+            tc = CMakeDeps(self)
+            tc.generate()
+        else: # Even with multiple toolchains (see below), can only have one "deps" generator as multiple ones will collide
+            tc = AutotoolsDeps(self)
+            tc.generate()
+
+        # Setup autotools on all platforms because we need to run autotools.configure when using CMake
+        tc = AutotoolsToolchain(self)
+        tc.configure_args.append("--enable-thread-safe")
+        tc.configure_args.append(f'--with-gmp={unix_path(self, self.dependencies[str(self.options.exact_int)].package_folder)}')
+        tc.configure_args.append(f'--enable-shared={"yes" if self.options.shared else "no"}')
+        tc.configure_args.append(f'--enable-static={"yes" if not self.options.shared else "no"}')
```

Comment:
It does not harm, but it's already managed by AutotoolsToolchain automatically: https://docs.conan.io/en/latest/reference/conanfile/tools/gnu/autotoolstoolchain.html#attributes

Reply:
Hi @uilianries,
Good point. I've removed the redundant lines.